base/inference.jl

```diff
@@ -2,14 +2,44 @@
 
 import Core: _apply, svec, apply_type, Builtin, IntrinsicFunction
 
-#### parameters limiting potentially-infinite types ####
-const MAX_TYPEUNION_LEN = 3
-const MAX_TYPE_DEPTH = 7
-const MAX_TUPLETYPE_LEN = 15
-const MAX_TUPLE_DEPTH = 4
-
-const MAX_TUPLE_SPLAT = 16
-const MAX_UNION_SPLITTING = 4
+immutable InferenceParams
+    # optimization
+    optimize::Bool
+    inlining::Bool
+    needtree::Bool
+    cached::Bool
+
+    # parameters limiting potentially-infinite types
+    MAX_TYPEUNION_LEN::Int
```

Comment:
I would just keep all of these as global constants unless you have a use case for changing them on some calls.